File: executioncontext/workspacecontext/src/test/java/org/jetbrains/bsp/bazel/workspacecontext/WorkspaceContextConstructorImplTest.java
package org.jetbrains.bsp.bazel.workspacecontext;

import static org.assertj.core.api.Assertions.assertThat;

import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
import io.vavr.collection.List;
import io.vavr.control.Option;
import io.vavr.control.Try;
import org.jetbrains.bsp.bazel.projectview.model.ProjectView;
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewTargetsSection;
import org.jetbrains.bsp.bazel.workspacecontext.entries.ExecutionContextTargetsEntity;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

public class WorkspaceContextConstructorImplTest {

  private WorkspaceContextConstructorImpl workspaceContextConstructor;

  @BeforeEach
  public void beforeEach() {
    // given
    this.workspaceContextConstructor = new WorkspaceContextConstructorImpl();
  }

  @Nested
  @DisplayName("Try<WorkspaceContext> construct(Try<ProjectView> projectViewTry) tests")
  class ConstructTryTest {

    @Test
    public void shouldReturnFailureIfProjectViewIsFailure() {
      // given
      var projectViewTry = Try.<ProjectView>failure(new Exception("exception message"));

      // when
      var workspaceContextTry = workspaceContextConstructor.construct(projectViewTry);

      // then
      assertThat(workspaceContextTry.isFailure()).isTrue();
      assertThat(workspaceContextTry.getCause().getClass()).isEqualTo(Exception.class);
      assertThat(workspaceContextTry.getCause().getMessage()).isEqualTo("exception message");
    }
  }

  @Nested
  @DisplayName("Try<WorkspaceContext> construct(ProjectView projectView) tests")
  class ConstructTest {

    @Test
    public void shouldReturnSuccessIfProjectViewIsValid() {
      // given
      var projectView =
          ProjectView.builder()
              .targets(
                  Option.of(
                      new ProjectViewTargetsSection(
                          List.of(
                              new BuildTargetIdentifier("//included_target1"),
                              new BuildTargetIdentifier("//included_target2"),
                              new BuildTargetIdentifier("//included_target3")),
                          List.of(new BuildTargetIdentifier("//excluded_target1")))))
              .build();

      // when
      var workspaceContextTry = workspaceContextConstructor.construct(projectView);

      // then
      assertThat(workspaceContextTry.isSuccess()).isTrue();
      var workspaceContext = workspaceContextTry.get();

      var expectedTargets =
          new ExecutionContextTargetsEntity(
              List.of(
                  new BuildTargetIdentifier("//included_target1"),
                  new BuildTargetIdentifier("//included_target2"),
                  new BuildTargetIdentifier("//included_target3")),
              List.of(new BuildTargetIdentifier("//excluded_target1")));
      assertThat(workspaceContext.getTargets()).isEqualTo(expectedTargets);
    }
  }
}

File: executioncontext/workspacecontext/src/test/java/org/jetbrains/bsp/bazel/workspacecontext/WorkspaceContextConstructorImplTest.kt
package org.jetbrains.bsp.bazel.workspacecontext

import ch.epfl.scala.bsp4j.BuildTargetIdentifier
import io.kotest.matchers.shouldBe
import io.vavr.control.Try
import io.vavr.collection.List
import org.jetbrains.bsp.bazel.projectview.model.ProjectView
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewTargetsSection
import org.jetbrains.bsp.bazel.workspacecontext.entries.ExecutionContextTargetsEntity
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class WorkspaceContextConstructorImplTest {

    private lateinit var workspaceContextConstructor: WorkspaceContextConstructorImpl

    @BeforeEach
    fun beforeEach() {
        // given
        this.workspaceContextConstructor = WorkspaceContextConstructorImpl()
    }

    @Nested
    @DisplayName("fun construct(projectViewTry: Try<ProjectView>): Try<WorkspaceContext> tests")
    inner class ConstructTryTest {

        @Test
        fun `should return failure if project view is failure`() {
            // given
            val projectViewTry = Try.failure<ProjectView>(Exception("exception message"))

            // when
            val workspaceContextTry = workspaceContextConstructor.construct(projectViewTry)

            // then
            workspaceContextTry.isFailure shouldBe true
            workspaceContextTry.cause::class shouldBe Exception::class
            workspaceContextTry.cause.message shouldBe "exception message"
        }
    }

    @Nested
    @DisplayName("fun construct(projectView: ProjectView): Try<WorkspaceContext> tests")
    inner class ConstructTest {

        @Test
        fun `should return success if project view is valid`() {
            // given
            val projectView =
                ProjectView.Builder(
                    targets =
                    ProjectViewTargetsSection(
                        List.of(
                            BuildTargetIdentifier("//included_target1"),
                            BuildTargetIdentifier("//included_target2"),
                            BuildTargetIdentifier("//included_target3")
                        ),
                        List.of(BuildTargetIdentifier("//excluded_target1")),
                    )
                ).build()

            // when
            val workspaceContextTry = workspaceContextConstructor.construct(projectView)

            // then
            workspaceContextTry.isSuccess shouldBe true
            val workspaceContext = workspaceContextTry.get()

            val expectedTargets =
                ExecutionContextTargetsEntity(
                    List.of(
                        BuildTargetIdentifier("//included_target1"),
                        BuildTargetIdentifier("//included_target2"),
                        BuildTargetIdentifier("//included_target3")
                    ), List.of(BuildTargetIdentifier("//excluded_target1"))
                )
            workspaceContext.targets shouldBe expectedTargets
        }
    }
}
